src/Sarif/FileEncodings.cs

```diff
@@ -0,0 +1,137 @@
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+    OTHER DEALINGS IN THE SOFTWARE. 
+
+    License: http://www.opensource.org/licenses/mit-license.php
+    Website: https://github.com/khalidsalomao/SimpleHelpers.Net
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class FileEncoding
+    {
+        /// <summary>
+        /// Detects if contains textual data.
+        /// </summary>
+        /// <param name="rawData">The raw data.</param>
+        public static bool CheckForTextualData(byte[] rawData, out Encoding encoding)
+        {
+            return CheckForTextualData(rawData, 0, rawData.Length, out encoding);
+        }
+
+        /// <summary>
+        /// Detects if contains textual data.
+        /// </summary>
+        /// <param name="rawData">The raw data.</param>
+        /// <param name="start">The start.</param>
+        /// <param name="count">The count.</param>
+        public static bool CheckForTextualData(byte[] rawData, int start, int count, out Encoding encoding)
+        {
+            encoding = null;
+
+            if (rawData.Length < count || count < 4 || start + 1 >= count)
+            {
+                return true;
+            }
+
+            if (CheckForByteOrderMark(rawData, out encoding, start))
+            {
+                return true;
+            }
+
+            // http://stackoverflow.com/questions/910873/how-can-i-determine-if-a-file-is-binary-or-text-in-c
+            // http://www.gnu.org/software/diffutils/manual/html_node/Binary.html
+            // count the number od null bytes sequences
+            // considering only sequences of 2 0s: "\0\0" or control characters below 10
+            int nullSequences = 0;
+            int controlSequences = 0;
+            for (int i = start + 1; i < count; i++)
+            {
+                if (rawData[i - 1] == 0 && rawData[i] == 0)
+                {
+                    if (++nullSequences > 1)
+                    {
+                        break;
+                    }
+                }
+                else if (rawData[i - 1] == 0 && rawData[i] < 10)
+                {
+                    ++controlSequences;
+                }
+            }
+
+            // is text if there is no null byte sequences or less than 10% of the buffer has control characters
+            return nullSequences == 0 && (controlSequences <= (rawData.Length / 10));
+        }
+
+        /// <summary>
+        /// Detects if data has bytes order mark to indicate its encoding for textual data.
+        /// </summary>
+        /// <param name="rawData">The raw data.</param>
+        /// <param name="start">The start.</param>
+        /// <returns></returns>
+        public static bool CheckForByteOrderMark(byte[] rawData, out Encoding encoding, int start = 0)
```

Comment:
>CheckForByteOrderMark [](http://example.com/codeflow?start=27&length=21)

We do not have explicit testing/use of BOM detection. I have a stubbed test but haven't nailed down how to generate useful test cases. I'm leaving this (and a stub test) for the future.

In practice, detecting/handling esoteric encodings is not a prominent scenario for most analysis scenarios.